### PR TITLE
[reland] Update singleton int to error when inequality relation is undefined

### PR DESCRIPTION
**Pull request by soulitzer**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #110728
* #110676
* #110675
* #110674
* #110673
* __->__ #110672

reland of https://github.com/pytorch/pytorch/pull/110044

### CODE REVIEW

**Comment by soulitzer** [comment]
The following line is changed in the reland to fix internal build failures:
https://github.com/pytorch/pytorch/pull/110672/commits/ba5fcb502aefeac205f9f5970a7f1cbfc6c864de

File: c10/core/SingletonSymNodeImpl.cpp [hunk 1]
```diff
@@ -1,3 +1,65 @@
...
+      if (*mb_si == *mb_si2) {
+        return true;
+      }
+      TORCH_CHECK(false, "Singleton int ", op, ": Relation is indeterminate");
+    }
+    if (rhs->constant_int() && *rhs->constant_int() <= 2) {
+      return true;
+    }
+    TORCH_CHECK(false, "Singleton int ", op, ": Relation is indeterminate");
+  } else if (rhs->singleton_int()) {
```
Comment (soulitzer) [inline c10/core/SingletonSymNodeImpl.cpp:25]: This is the only change in the reland.

Previously:
```cpp
stderr: xplat/caffe2/c10/core/SingletonSymNodeImpl.cpp:25:19: error: declaration shadows a local variable [-Werror,-Wshadow]
  } else if (auto mb_si = rhs->singleton_int()) {
                  ^
xplat/caffe2/c10/core/SingletonSymNodeImpl.cpp:14:12: note: previous declaration is here
  if (auto mb_si = lhs->singleton_int()) {
           ^
1 error generated.
```

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**